ir_datasets/datamaestro/datasets.py

```diff
@@ -0,0 +1,210 @@
+
+    def _prepare(self, download=False) -> AdhocDocuments:
+        return AdhocAssessments(id=self.fullid)
+
+
+class Queries(Dataset):
+    SUFFIX = "queries"
+    configtype = AdhocTopics
+
+    def download(self, force=False):
+        # Triggers download
+        self.irds_ds.queries_iter()
+        return True
+
+    def _prepare(self, download=False) -> AdhocDocuments:
+        return AdhocTopics(id=self.fullid)
+
+
+# class ScoredDocuments(Dataset):
+#     SUFFIX = "scored-documents"
+#     base = ScoredDocuments
+
+
+class Documents(Dataset):
+    SUFFIX = "documents"
+    configtype = AdhocDocuments
+
+    def download(self, force=False):
+        # Triggers download
+        self.irds_ds.docs_iter()
+        return True
+
+    def _prepare(self, download=False) -> AdhocDocuments:
+        return AdhocDocuments(id=self.fullid)
+
+    def download(self, force=False):
+        self.irds_ds.docs_path()
+        return True
+
+
+class TrainingTripletsDataset(Dataset):
+    SUFFIX = "docpairs"
+
+    @property
+    def configtype(self):
+        if issubclass(self.irds_ds.docpairs_cls(), GenericDocPair):
+            return TrainingTripletsLines
+
+    def _prepare(self, download=False) -> AdhocDocuments:
+        if issubclass(self.irds_ds.docpairs_cls(), GenericDocPair):
+            return TrainingTripletsLines(
+                id=self.fullid,
+                ids=True,
+                path=Path(self.irds_ds.docpairs_path()),
+            )
+        raise NotImplementedError("For %s", self.irds_ds.docpairs_cls())
+
+
+
+class AdhocRunDataset(Dataset):
+    SUFFIX = "scoreddocs"
+    base = AdhocRun
+    configtype = AdhocRun
+
+    def _prepare(self, download=False) -> AdhocDocuments:
+        return AdhocRun(id=self.fullid)
+
+
+class Collection(Dataset):
+    base = Adhoc
+    assessements: Qrels
+    topics: Queries
+
+    # FIXME: find a proper way to get the collection path
+    # @property
+    # def datapath(self):
+    #     return Path(self.irds_ds._dlc._path)
+
+    # def hasfiles(self):
+    #     print(self.irds_id)
+    #     return self.irds_ds._dlc._path is not None
+
+    def _prepare(self, download=False) -> AdhocDocuments:
+        return Adhoc(
+            id=self.fullid,
+            topics=self.topics.prepare(download),
+            assessments=self.assessments.prepare(download),
+            documents=self.documents.prepare(download),
+        )
+
+
+class Datasets:
+    """Simple wrapper holding related data pieces"""
+
+    def __init__(self, key, description):
+        self.id = key
+        self.description = description
+        self.datasets = []
+
+    def __iter__(self) -> Iterable[AbstractDataset]:
+        return self.datasets.__iter__()
+
+
+def build(repository):
+    """Builds a repository by using ir_datasets registry"""
+    datasets = {}
+    bykey = {}
+
+    def add(cid, ds):
+        datasets[cid].datasets.append(ds)
+        bykey[ds.id] = ds
+
+    for dataset_id in registry:
```

Comment:
Note that not all datasets are added directly to the registry -- CLIRMatrix has a ton of combinations so it didn't make sense to add them all. Instead, a pattern-matching approach is used to build the dataset on demand. See an example [here](https://github.com/allenai/ir_datasets/blob/master/examples/clirmatrix_example.py).

Also, there's the potential for "extension" datasets (though afaik they are not really used yet). The idea is that if you have your own dataset that you do not necessarily want to add to the public registry, you can have a package that adds it. E.g., the example [here](https://github.com/seanmacavaney/dummy-irds-ext). The trick would be to make sure that the extension is imported before the registry is built. We've also talked about how to auto-import extensions (e.g. via an environment variable and/or a file that defines them in ~/.ir_datasets/).

I'm not sure what the best way to handle these cases would be. I suspect it would be to convert the datamaestro-style ID to an IRDS ID in `search` above and pass it to `ir_datasets.load`, rather than building up a list ahead of time. PyTerrier takes the approach of both [pre-loading the ones in the registry](https://github.com/terrier-org/pyterrier/blob/master/pyterrier/datasets.py#L977) and [allowing them to be built on the fly](https://github.com/terrier-org/pyterrier/blob/master/pyterrier/datasets.py#L994).

Reply:
OK. The natural in datamaestro way would be to provide datasets for queries, documents and qrels since they are limited in number, and let the user build any combination of those.

Reply:
Even with that arrangement, there are 153,819 individual units for CLIRMatrix, which is a tad unmanageable.

Reply:
The full list of source files is here: http://www.cs.jhu.edu/~shuosun/clirmatrix/data/downloads.json.gz